### PR TITLE
Dem/random variable utility

### PR DESCRIPTION
**Pull request by GuillermoCasas**
**Description**
This PR introduces the object RandomVariable that can be used to obtain samples from the chosen distribution as well as other associated tools (mean, etc.) which are largely still to be developed.

A first intance of the abstract class is introduced to generate real numbers out of an arbitrary piecewise linear distribution.

The idea is to use this utility to manage random-number generation in DEMApplication (for the moment).

**Changelog**
- Added RandomVariable object
- Added PiecewiseLinearRandomVariable object
- Created Test that iterates by repeatedly sampling from the piecewise linear RV ontil the recovered empirical distribution is close enough to the input distribution. Also the mean calculation is checked.

### CODE REVIEW

**Comment by rubenzorrilla** [comment]
Maybe it would be nice to have this contribution in the  `StatisticsApplication` so anyone can use it.

**Comment by GuillermoCasas** [comment]
> Maybe it would be nice to have this contribution in the `StatisticsApplication` so anyone can use it.

Sure, but I think it is better to test it for a while in DEMApplication

File: applications/DEMApplication/custom_utilities/piecewise_linear_random_variable.h [hunk 1]
```diff
@@ -0,0 +1,94 @@
+//
+// Authors:
+// Guillermo Casas gcasas@cimne-upc.edu
+//
+//README::::look to the key word "VERSION" if you want to find all the points where you have to change something so that you can pass from a kdtree to a bin data search structure;
```
Comment (maceligueta) [inline applications/DEMApplication/custom_utilities/piecewise_linear_random_variable.h]: VERSION?

File: applications/DEMApplication/custom_utilities/random_variable.h [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    KRATOS_CLASS_POINTER_DEFINITION(RandomVariable);
+
+    /// Default constructor
+    RandomVariable();
+    RandomVariable(const Parameters rParameters);
+
+    /// Destructor
+    ~RandomVariable(){};
+
+    virtual double Sample(){KRATOS_ERROR << "You are calling the 'Sample' function of the abstract class 'RandomVariable'. Please instantiate a specific derived class instead."; return 0.0;};
```
Comment (maceligueta) [inline applications/DEMApplication/custom_utilities/random_variable.h:34]: return 0.0 will never happen, as the KRATOS_ERROR stops the computation

File: applications/DEMApplication/custom_utilities/random_variable.h [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    virtual double GetMean(){KRATOS_ERROR << "You are calling 'GetMean' function of the abstract class 'RandomVariable'. Please instantiate a specific derived class instead."; return 0.0;};
```
Comment (maceligueta) [inline applications/DEMApplication/custom_utilities/random_variable.h:36]: same

File: applications/DEMApplication/custom_utilities/random_variable.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
+//
+// Authors:
+// Guillermo Casas gcasas@cimne-upc.edu
+//
+//README::::look to the key word "VERSION" if you want to find all the points where you have to change something so that you can pass from a kdtree to a bin data search structure;
```
Comment (maceligueta) [inline applications/DEMApplication/custom_utilities/random_variable.h]: 'VERSION' comment

File: applications/DEMApplication/custom_utilities/piecewise_linear_random_variable.h (hunk 1 above)
Comment (GuillermoCasas) [reply applications/DEMApplication/custom_utilities/piecewise_linear_random_variable.h]: I removed it.

**Comment by maceligueta** [comment]
Maybe you'd like to add your authorship in some of the files you added which don't have it.

File: applications/DEMApplication/custom_utilities/random_variable.h (hunk 3 above)
Comment (maceligueta) [reply applications/DEMApplication/custom_utilities/random_variable.h:36]: oh, maybe it gives a warning if you don't put it. Nevermind

**Comment by GuillermoCasas** [comment]
> Maybe you'd like to add your authorship in some of the files you added which don't have it.

I just did.

**Review by maceligueta — APPROVED** [review]

**Pull request merged**